Redesigning genericClient

abhijeetmanhas opened the pull request:
<!--
We know that working on sunpy and submitting pull requests takes effort, and we appreciate your time.
Thank you.

Please be aware that everyone has to follow our code of conduct:
https://github.com/sunpy/sunpy/blob/master/CODE_OF_CONDUCT.rst

Also these comments are hidden when you submit this github pull request.

We have several automatic features that run when a pull request is open.
They can appear daunting but do not worry about them!
Here is a brief explanation of them: https://docs.sunpy.org/en/latest/dev_guide/pr_review_procedure.html#continuous-integration.
-->

### Description
<!--
Provide a general description of what your pull request does.

If the pull request closes any open issues you can add this.
If you replace <Issue Number> with a number GitHub will automatically link it.
If it doesn't, please remove the following line.
-->

Key highlights:

- Easy to define a new "simple" Fido client. Only define `pattern`, `baseurl`, and `register_values()`. This further uses scraper and `~sunpy.extern.parse` for extracting metadata.
- More metadata is shown in the response tables. Like `SatelliteNumber`, `Level`, `Detector`, `Resolution`,  `Wavelength` will be shown which earlier were either not present or `np.nan`.
- Nothing is default Attr for any search made to client. So if client supports 3 levels, and we don't specify level in the query, data fo all levels shall be shown.
- Refactoring in `GenericClient`, removed/generalized/updated a lot of methods.

Fixes #3715, Fixes #3337, Fixes  #3321, Fixes #3306, Fixes #2314.

pep8speaks commented:
Hello @abhijeetmanhas! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`sunpy/extern/parse.py`](https://github.com/sunpy/sunpy/blob/f5a41b2f1357ff0c2c691c9cd7ddff429e186037/sunpy/extern/parse.py):

> [Line 533:11](https://github.com/sunpy/sunpy/blob/f5a41b2f1357ff0c2c691c9cd7ddff429e186037/sunpy/extern/parse.py#L533): [E114](https://duckduckgo.com/?q=pep8%20E114) indentation is not a multiple of four (comment)
> [Line 534:11](https://github.com/sunpy/sunpy/blob/f5a41b2f1357ff0c2c691c9cd7ddff429e186037/sunpy/extern/parse.py#L534): [E111](https://duckduckgo.com/?q=pep8%20E111) indentation is not a multiple of four
> [Line 536:11](https://github.com/sunpy/sunpy/blob/f5a41b2f1357ff0c2c691c9cd7ddff429e186037/sunpy/extern/parse.py#L536): [E114](https://duckduckgo.com/?q=pep8%20E114) indentation is not a multiple of four (comment)
> [Line 537:11](https://github.com/sunpy/sunpy/blob/f5a41b2f1357ff0c2c691c9cd7ddff429e186037/sunpy/extern/parse.py#L537): [E111](https://duckduckgo.com/?q=pep8%20E111) indentation is not a multiple of four
> [Line 538:15](https://github.com/sunpy/sunpy/blob/f5a41b2f1357ff0c2c691c9cd7ddff429e186037/sunpy/extern/parse.py#L538): [E111](https://duckduckgo.com/?q=pep8%20E111) indentation is not a multiple of four
> [Line 539:19](https://github.com/sunpy/sunpy/blob/f5a41b2f1357ff0c2c691c9cd7ddff429e186037/sunpy/extern/parse.py#L539): [E111](https://duckduckgo.com/?q=pep8%20E111) indentation is not a multiple of four
> [Line 540:15](https://github.com/sunpy/sunpy/blob/f5a41b2f1357ff0c2c691c9cd7ddff429e186037/sunpy/extern/parse.py#L540): [E111](https://duckduckgo.com/?q=pep8%20E111) indentation is not a multiple of four
> [Line 541:19](https://github.com/sunpy/sunpy/blob/f5a41b2f1357ff0c2c691c9cd7ddff429e186037/sunpy/extern/parse.py#L541): [E111](https://duckduckgo.com/?q=pep8%20E111) indentation is not a multiple of four
> [Line 542:15](https://github.com/sunpy/sunpy/blob/f5a41b2f1357ff0c2c691c9cd7ddff429e186037/sunpy/extern/parse.py#L542): [E111](https://duckduckgo.com/?q=pep8%20E111) indentation is not a multiple of four
> [Line 543:19](https://github.com/sunpy/sunpy/blob/f5a41b2f1357ff0c2c691c9cd7ddff429e186037/sunpy/extern/parse.py#L543): [E111](https://duckduckgo.com/?q=pep8%20E111) indentation is not a multiple of four
> [Line 566:13](https://github.com/sunpy/sunpy/blob/f5a41b2f1357ff0c2c691c9cd7ddff429e186037/sunpy/extern/parse.py#L566): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 606:9](https://github.com/sunpy/sunpy/blob/f5a41b2f1357ff0c2c691c9cd7ddff429e186037/sunpy/extern/parse.py#L606): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 613:10](https://github.com/sunpy/sunpy/blob/f5a41b2f1357ff0c2c691c9cd7ddff429e186037/sunpy/extern/parse.py#L613): [E225](https://duckduckgo.com/?q=pep8%20E225) missing whitespace around operator
> [Line 614:10](https://github.com/sunpy/sunpy/blob/f5a41b2f1357ff0c2c691c9cd7ddff429e186037/sunpy/extern/parse.py#L614): [E225](https://duckduckgo.com/?q=pep8%20E225) missing whitespace around operator
> [Line 615:10](https://github.com/sunpy/sunpy/blob/f5a41b2f1357ff0c2c691c9cd7ddff429e186037/sunpy/extern/parse.py#L615): [E225](https://duckduckgo.com/?q=pep8%20E225) missing whitespace around operator
> [Line 707:5](https://github.com/sunpy/sunpy/blob/f5a41b2f1357ff0c2c691c9cd7ddff429e186037/sunpy/extern/parse.py#L707): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 793:17](https://github.com/sunpy/sunpy/blob/f5a41b2f1357ff0c2c691c9cd7ddff429e186037/sunpy/extern/parse.py#L793): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 806:25](https://github.com/sunpy/sunpy/blob/f5a41b2f1357ff0c2c691c9cd7ddff429e186037/sunpy/extern/parse.py#L806): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 820:25](https://github.com/sunpy/sunpy/blob/f5a41b2f1357ff0c2c691c9cd7ddff429e186037/sunpy/extern/parse.py#L820): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 823:21](https://github.com/sunpy/sunpy/blob/f5a41b2f1357ff0c2c691c9cd7ddff429e186037/sunpy/extern/parse.py#L823): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 824:21](https://github.com/sunpy/sunpy/blob/f5a41b2f1357ff0c2c691c9cd7ddff429e186037/sunpy/extern/parse.py#L824): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 890:39](https://github.com/sunpy/sunpy/blob/f5a41b2f1357ff0c2c691c9cd7ddff429e186037/sunpy/extern/parse.py#L890): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 924:13](https://github.com/sunpy/sunpy/blob/f5a41b2f1357ff0c2c691c9cd7ddff429e186037/sunpy/extern/parse.py#L924): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 986:25](https://github.com/sunpy/sunpy/blob/f5a41b2f1357ff0c2c691c9cd7ddff429e186037/sunpy/extern/parse.py#L986): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 987:25](https://github.com/sunpy/sunpy/blob/f5a41b2f1357ff0c2c691c9cd7ddff429e186037/sunpy/extern/parse.py#L987): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 1064:101](https://github.com/sunpy/sunpy/blob/f5a41b2f1357ff0c2c691c9cd7ddff429e186037/sunpy/extern/parse.py#L1064): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (105 > 100 characters)
> [Line 1065:58](https://github.com/sunpy/sunpy/blob/f5a41b2f1357ff0c2c691c9cd7ddff429e186037/sunpy/extern/parse.py#L1065): [E261](https://duckduckgo.com/?q=pep8%20E261) at least two spaces before inline comment
> [Line 1065:101](https://github.com/sunpy/sunpy/blob/f5a41b2f1357ff0c2c691c9cd7ddff429e186037/sunpy/extern/parse.py#L1065): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (115 > 100 characters)
> [Line 1071:17](https://github.com/sunpy/sunpy/blob/f5a41b2f1357ff0c2c691c9cd7ddff429e186037/sunpy/extern/parse.py#L1071): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 1078:17](https://github.com/sunpy/sunpy/blob/f5a41b2f1357ff0c2c691c9cd7ddff429e186037/sunpy/extern/parse.py#L1078): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 1085:17](https://github.com/sunpy/sunpy/blob/f5a41b2f1357ff0c2c691c9cd7ddff429e186037/sunpy/extern/parse.py#L1085): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 1090:17](https://github.com/sunpy/sunpy/blob/f5a41b2f1357ff0c2c691c9cd7ddff429e186037/sunpy/extern/parse.py#L1090): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 1093:17](https://github.com/sunpy/sunpy/blob/f5a41b2f1357ff0c2c691c9cd7ddff429e186037/sunpy/extern/parse.py#L1093): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 1098:17](https://github.com/sunpy/sunpy/blob/f5a41b2f1357ff0c2c691c9cd7ddff429e186037/sunpy/extern/parse.py#L1098): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 1101:17](https://github.com/sunpy/sunpy/blob/f5a41b2f1357ff0c2c691c9cd7ddff429e186037/sunpy/extern/parse.py#L1101): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 1108:17](https://github.com/sunpy/sunpy/blob/f5a41b2f1357ff0c2c691c9cd7ddff429e186037/sunpy/extern/parse.py#L1108): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 1114:17](https://github.com/sunpy/sunpy/blob/f5a41b2f1357ff0c2c691c9cd7ddff429e186037/sunpy/extern/parse.py#L1114): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 1120:17](https://github.com/sunpy/sunpy/blob/f5a41b2f1357ff0c2c691c9cd7ddff429e186037/sunpy/extern/parse.py#L1120): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 1204:13](https://github.com/sunpy/sunpy/blob/f5a41b2f1357ff0c2c691c9cd7ddff429e186037/sunpy/extern/parse.py#L1204): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 1286:9](https://github.com/sunpy/sunpy/blob/f5a41b2f1357ff0c2c691c9cd7ddff429e186037/sunpy/extern/parse.py#L1286): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 1320:9](https://github.com/sunpy/sunpy/blob/f5a41b2f1357ff0c2c691c9cd7ddff429e186037/sunpy/extern/parse.py#L1320): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 1347:101](https://github.com/sunpy/sunpy/blob/f5a41b2f1357ff0c2c691c9cd7ddff429e186037/sunpy/extern/parse.py#L1347): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (112 > 100 characters)





* In the file [`sunpy/net/dataretriever/sources/fermi_gbm.py`](https://github.com/sunpy/sunpy/blob/f5a41b2f1357ff0c2c691c9cd7ddff429e186037/sunpy/net/dataretriever/sources/fermi_gbm.py):

> [Line 51:101](https://github.com/sunpy/sunpy/blob/f5a41b2f1357ff0c2c691c9cd7ddff429e186037/sunpy/net/dataretriever/sources/fermi_gbm.py#L51): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (123 > 100 characters)
> [Line 52:101](https://github.com/sunpy/sunpy/blob/f5a41b2f1357ff0c2c691c9cd7ddff429e186037/sunpy/net/dataretriever/sources/fermi_gbm.py#L52): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (102 > 100 characters)

abhijeetmanhas commented:
### Tasks
 * [ ] Generalize `fermi`, `eve`, `lyra`
 * [ ] Generalize `goes`, `noaa`, `norh`
 * [ ] Move `rhessi` out of Generic
 * [ ] Update all tests and docs

Cadair commented on sunpy/net/_attrs.py:
```diff
@@ -243,3 +243,13 @@ class Physobs(SimpleAttr):
     """
+
+
+class Source(SimpleAttr):
```
aren't these in VSO at the moment?

abhijeetmanhas replied:
Moved from vso to net

Cadair commented on sunpy/net/dataretriever/client.py:
```diff
@@ -89,10 +88,17 @@ class GenericClient(BaseClient):
 
     @classmethod
-    def _get_patterns_for_attrs(cls, **kwargs):
+    def pre_hook(cls, *args, **kwargs):
```
What are the args here? Does this need to be a full `*args, **kwargs`?

I think the query needs to be passed in here? Does anything else?

abhijeetmanhas replied:
Yes, `*args` and `**kwargs` of query shall be used for translation here. This pre-hook will be overwritten for some clients like `norh`. Where wavelength shall translate to either `tcz` or `tca` in the url.

Cadair commented on sunpy/net/dataretriever/client.py:
```diff
@@ -225,7 +107,34 @@ def _can_handle_query(cls, *query):
-        raise NotImplementedError
+        from sunpy.net import attrs as a
+
+        required = {a.Time, a.Instrument}
```
This needs to be a class attr?

abhijeetmanhas replied:
For most clients, `Instrument` and `Time` are required attrs so the method is same for all clients.
If any client deviates, it will be implemented in the  subclass.

Cadair replied:
yeah but they shouldn't need to override this method to set this, they should just be able to set a class attr?

nabobalis replied:
This is still in the method and not a class attr?

abhijeetmanhas replied:
yes, since it is by default `instrument` and `Time`. If any client want to overwrite it, then it can be defined as class attributes for that clients, and the generic `_can_handle_query` checks if they are defined as class attributes or not. If not, it will follow the default rule.

Cadair replied:
But why not just set required on this class, so it's always defined and then use it consistently in this method rather than doing `hasattr`. I would also think it would be a good idea to define `optional = set()` as a class attr as well, so these are always defined and we don't have to check if they are there.

abhijeetmanhas replied:
Done

nabobalis commented on sunpy/net/_attrs.py:
```diff
@@ -243,3 +243,13 @@ class Physobs(SimpleAttr):
     """
+
+
+class Source(SimpleAttr):
```
Are these not defined under the VSO namespace?

abhijeetmanhas replied:
Moved from vso to net

Cadair commented on sunpy/net/dataretriever/client.py:
```diff
@@ -225,7 +107,34 @@ def _can_handle_query(cls, *query):
+
+        required = {a.Time, a.Instrument}
+        adict = cls.register_values()
+        optional = {i for i in adict.keys()}
```
This needs to be everything in register_values that isn't in required?
```suggestion
        optional = {i for i in adict.keys()}difference(required)
```

abhijeetmanhas replied:
yeah, True will change it accordingly

nabobalis commented:
Should be rebased on top of master.

Cadair commented on sunpy/net/dataretriever/client.py:
```diff
@@ -225,7 +107,34 @@ def _can_handle_query(cls, *query):
+        map_ = OrderedDict()
+        start = parse_time("{}/{}/{}".format(exdict['year'], exdict['month'], exdict['day']))
+        end = start + TimeDelta(1 * u.day - 1 * u.millisecond)
+        map_['Start Time'] = start.strftime(TIME_FORMAT)
```
Why are we converting these to strings to put them in the response, can't we just pass astropy time objects to Table?

abhijeetmanhas replied:
Done

Cadair commented on sunpy/net/dataretriever/client.py:
```diff
@@ -137,86 +85,20 @@ class GenericClient(BaseClient):
-
     @classmethod
-    def _get_url_for_timerange(cls, timerange, **kwargs):
+    def pre_hook(cls, *args, **kwargs):
```
```suggestion
    def pre_search_hook(cls, *args, **kwargs):
```

abhijeetmanhas replied:
Done

Cadair commented on sunpy/net/dataretriever/client.py:
```diff
@@ -264,35 +173,27 @@ def _get_full_filenames(self, qres, filenames, path):
+        filesmeta = scraper._extract_files_meta(timerange, extractor=pattern, matcher=matchdict)
+        metalist = []
+        for i in filesmeta:
+            map_ = self.post_hook(i, matchdict)
```
What do you think the pros/cons of putting this loop inside `post_hook` are? Are there cases where the source would want to handle all the returns as a group? i.e. to filter them or something mad?

abhijeetmanhas replied:
Passing large `filesmeta`list of dictionaries which may contain a lot of metada to a method will increase memory usage (I'm not very certain about it).
There are no filtering left, all that is handled by the `matchdict` passed in scraper, as we decided in the last meeting.
The `post_hook` will be overwritten for clients like `norh` to convert `tcz` or `tca` in the url to `17*u.GHz` or `34 *u.GHz`.

Cadair commented on sunpy/net/dataretriever/sources/noaa.py:
```diff
@@ -200,6 +149,22 @@ def _get_url_for_timerange(self, timerange, **kwargs):
+        extractor2 = '{}/warehouse/{year:4d}/{}'
+        matchdict = {'Instrument': ['SOON'], 'Source': ['SWPC'], 'Provider': ['NOAA'], 'Physobs': ['SRS']}
+        metalist = []
+        for elem in args:
+            if isinstance(elem, a.Time):
+                timerange = TimeRange(elem.start, elem.end)
```
Can you not replace this with a call to `pre_hook`?

abhijeetmanhas replied:
Done

Cadair commented on sunpy/net/dataretriever/sources/noaa.py:
```diff
@@ -200,6 +149,22 @@ def _get_url_for_timerange(self, timerange, **kwargs):
+    def search(self, *args, **kwargs):
+        extractor1 = '{}/warehouse/{:4d}/SRS/{year:4d}{month:2d}{day:2d}SRS.txt'
+        extractor2 = '{}/warehouse/{year:4d}/{}'
+        matchdict = {'Instrument': ['SOON'], 'Source': ['SWPC'], 'Provider': ['NOAA'], 'Physobs': ['SRS']}
```
Also this would come out of `pre_hook` right?

abhijeetmanhas replied:
Done

Cadair commented on sunpy/net/dataretriever/sources/norh.py:
```diff
@@ -35,119 +34,55 @@ class NoRHClient(GenericClient):
-        self.map_['physobs'] = ''
+    baseurl = r'ftp://solar-pub.nao.ac.jp/pub/nsro/norh/data/tcx/%Y/%m/(\w){3}%y%m%d'
+    pattern = '{}/tcx/{year:4d}/{month:2d}/{Wavelength:3l}{:4d}{day:2d}'
+    optional = {a.Wavelength, a.Source, a.Provider}
```
I thought the plan was to have `required` as a class attribute and derive optional from `register_attrs`?

abhijeetmanhas replied:
This can't be done, since we can't derive `wavelength` from register_attrs as we planned not to register it.

Cadair commented on sunpy/net/dataretriever/client.py:
```diff
@@ -137,86 +85,30 @@ class GenericClient(BaseClient):
-    does this through the `map\\_` dictionary which represents the
-    `~sunpy.net.attrs` objects as a dictionary.
     """
 
```
```suggestion
    baseurl = None
    pattern = None
```

abhijeetmanhas replied:
Done

nabobalis commented on sunpy/net/_attrs.py:
```diff
@@ -243,3 +243,37 @@ class Physobs(SimpleAttr):
+
+class Provider(SimpleAttr):
+    """
+    Specifies the data provider to search for data using Fido clients.
```
```suggestion
    Specifies the data provider to search for data using Fido.
```

nabobalis commented on sunpy/net/_attrs.py:
```diff
@@ -243,3 +243,37 @@ class Physobs(SimpleAttr):
+    Parameters
+    ----------
+    value : str
+
```
This should be completed.

abhijeetmanhas replied:
Done

nabobalis commented on sunpy/net/_attrs.py:
```diff
@@ -243,3 +243,37 @@ class Physobs(SimpleAttr):
+
+    Notes
+    -----
+    For VSO, More information about each provider may be found within in the VSO Registry.
```
```suggestion
    For VSO, more information about each provider may be found within in the VSO Registry.
```

nabobalis commented on sunpy/net/_attrs.py:
```diff
@@ -243,3 +243,37 @@ class Physobs(SimpleAttr):
+    -----
+    For VSO, More information about each provider may be found within in the VSO Registry.
+    For a list of VSO providers see
+    https://sdac.virtualsolar.org/cgi/show_details?keyword=PROVIDER.
```
```suggestion
```

nabobalis commented on sunpy/net/_attrs.py:
```diff
@@ -243,3 +243,37 @@ class Physobs(SimpleAttr):
+    Notes
+    -----
+    For VSO, More information about each provider may be found within in the VSO Registry.
+    For a list of VSO providers see
```
```suggestion
    See `VSO providers <https://sdac.virtualsolar.org/cgi/show_details?keyword=PROVIDER>`__.
```

nabobalis commented on sunpy/net/_attrs.py:
```diff
@@ -243,3 +243,37 @@ class Physobs(SimpleAttr):
+
+class Source(SimpleAttr):
+    """
+    Data sources that Fido clients can search on.
```
```suggestion
    Data sources that Fido can search with.
```

nabobalis commented on sunpy/net/_attrs.py:
```diff
@@ -243,3 +243,37 @@ class Physobs(SimpleAttr):
+
+    Parameters
+    ----------
+    value : str
```
This should be completed.

abhijeetmanhas replied:
Done

nabobalis commented on sunpy/net/_attrs.py:
```diff
@@ -243,3 +243,37 @@ class Physobs(SimpleAttr):
+
+    Notes
+    -----
+    For VSO, More information about each source may be found within in the VSO Registry.
```
```suggestion
    For VSO, more information about each source may be found within in the VSO Registry.
```

nabobalis commented on sunpy/net/_attrs.py:
```diff
@@ -243,3 +243,37 @@ class Physobs(SimpleAttr):
+    Notes
+    -----
+    For VSO, More information about each source may be found within in the VSO Registry.
+    User Interface programmers should note that some names may be encoded as
```
```suggestion
    See `VSO sources <https://sdac.virtualsolar.org/cgi/show_details?keyword=SOURCE>`__.
```

nabobalis commented on sunpy/net/_attrs.py:
```diff
@@ -243,3 +243,37 @@ class Physobs(SimpleAttr):
+    -----
+    For VSO, More information about each source may be found within in the VSO Registry.
+    User Interface programmers should note that some names may be encoded as
+    UTF-8. Please note that 'Source' is used internally by VSO to represent
```
```suggestion
    Please note that 'Source' is used internally by VSO to represent
```

nabobalis commented on sunpy/net/_attrs.py:
```diff
@@ -243,3 +243,37 @@ class Physobs(SimpleAttr):
+    For VSO, More information about each source may be found within in the VSO Registry.
+    User Interface programmers should note that some names may be encoded as
+    UTF-8. Please note that 'Source' is used internally by VSO to represent
+    what the VSO Data Model refers to as 'Observatory'.  For a list of VSO sources
```
```suggestion
    what the VSO Data Model refers to as 'Observatory'.
```

nabobalis commented on sunpy/net/_attrs.py:
```diff
@@ -243,3 +243,37 @@ class Physobs(SimpleAttr):
+    User Interface programmers should note that some names may be encoded as
+    UTF-8. Please note that 'Source' is used internally by VSO to represent
+    what the VSO Data Model refers to as 'Observatory'.  For a list of VSO sources
+    see https://sdac.virtualsolar.org/cgi/show_details?keyword=SOURCE.
```
```suggestion
```

nabobalis commented on sunpy/net/vso/attrs.py:
```diff
@@ -64,40 +63,6 @@ def collides(self, other):
         return isinstance(other, self.__class__)
 
 
-class Provider(_attr.SimpleAttr):
```
You need to add these to the list of removed attrs at the bottom of this file.

abhijeetmanhas replied:
Done

nabobalis commented on sunpy/net/dataretriever/client.py:
```diff
@@ -137,86 +93,59 @@ class GenericClient(BaseClient):
         \\*args: `tuple`
-            The query attributes.
+            `sunpy.net.attrs` objects representing the query.
 
```
```suggestion
```

nabobalis commented on sunpy/net/dataretriever/client.py:
```diff
@@ -281,18 +257,22 @@ def search(self, *args, **kwargs):
         ----------
         \\*args: `tuple`
             `sunpy.net.attrs` objects representing the query.
+
```
```suggestion
```

nabobalis commented on sunpy/net/dataretriever/sources/fermi_gbm.py:
```diff
@@ -37,118 +36,28 @@ class GBMClient(GenericClient):
             attrs.Resolution: [
-            ("CSPEC", "CSPEC 128 channel spectra every 4.096 seconds."),
-            ("CTIME", "CTIME provides 8 channel spectra every 0.256 seconds")]
+            ("cspec", "CSPEC 128 channel spectra every 4.096 seconds."),
```
I feel like the attrs should be case insensitive, if they are not, we might want to change that.

abhijeetmanhas replied:
That would have been ideal, but the problem is, the arhive website may contain these attrs in any case, and our pattern matching will fail if we don't follow the website.

If we want to ignore case, that can also be done, but we have to then overwrite `pre_search_hook()` method (which norh does for converting 34 ghz to `tca`). But defining pre-hook for converting `CTIME` to `ctime` or `Ctime` looked like overkill.
So I defined the case the archive follows in register values itself.

nabobalis commented on sunpy/net/dataretriever/sources/goes.py:
```diff
@@ -223,183 +85,86 @@ class SUVIClient(GenericClient):
+                 '{:1d}_e{:7d}{ehour:2d}{eminute:2d}{esecond:2d}{:1d}_{}')
+    baseurl2 = (r'https://data.ngdc.noaa.gov/platforms/solar-space-observing-satellites/goes/goes{SatelliteNumber}/'
+                r'l2/data/suvi-l2-ci{wave:03}/%Y/%m/%d/dr_suvi-l2-ci{wave:03}_g{SatelliteNumber}_s%Y%m%dT%H%M%SZ_.*\.fits')
+    pattern2 = ('{}/goes/goes{SatelliteNumber:2d}/{}/dr_suvi-l{Level}-ci{Wavelength:03d}_g{SatelliteNumber:2d}_s'
```
In theory this could be reduced if the scrapper was smarter?

abhijeetmanhas replied:
Yeah, only convertors need to befined. (to convert level 2 to `/data/l2/` in baseurl). We can do everything with one baseurl itself.

abhijeetmanhas commented:
`fido_factory` tests updated. `test_fermi.py` can be reviewed as an example how all other sources clients test would look like.
Have to yet update/add tests for other clients, and look for failures outside net.

nabobalis commented:
Having had a quick glance over the last two commits, I see no reason not to carry on.

abhijeetmanhas commented:
Database tests only remaining to be updated.

nabobalis commented on sunpy/net/dataretriever/sources/tests/test_lyra_ud.py:
```diff
@@ -19,21 +22,22 @@ def LCClient():
 @pytest.mark.remote_data
 @pytest.mark.parametrize("timerange,url_start,url_end", [
-    (TimeRange('2012/1/7', '2012/1/7'),
+    (Time('2012/1/7', '2012/1/7'),
```
Does this return a timerange?

abhijeetmanhas replied:
It was `from sunpy.net.attrs import Time` So I just used it.

nabobalis replied:
Fair enough.

Cadair commented on docs/guide/acquiring_data/fido.rst:
```diff
@@ -62,19 +62,6 @@ Using ``Instrument`` as the first example, if you print the object::
 You get a full list of known values, a description and what "Clients" support those values (if you want to use a specific data source).
 This is supported for most attributes including the client specific ones.
 
-For VSO::
```
Are we really out of VSO specific attrs we register values for?

abhijeetmanhas replied:
Yes, though there are `Pscale`, `Extent`, `Pixel` etc, but nothing is registered for them.

Cadair commented on sunpy/net/dataretriever/client.py:
```diff
@@ -225,7 +176,63 @@ def _can_handle_query(cls, *query):
+        required = {a.Time, a.Instrument}
+        adict = cls.register_values()
+        optional = {i for i in adict.keys()} - required
+        if hasattr(cls, 'required'):
+            required = cls.required
+        if hasattr(cls, 'optional'):
+            optional = cls.optional
+        if not cls.check_attr_types_in_query(query, required, optional):
```
As per the thread above:
```suggestion
        if not cls.check_attr_types_in_query(query, cls.required, cls.optional):
```

abhijeetmanhas replied:
Done

Cadair commented on sunpy/net/dataretriever/client.py:
```diff
@@ -225,7 +176,63 @@ def _can_handle_query(cls, *query):
+
+    def post_search_hook(self, exdict, matchdict):
+        """
+        Helper function used after `search()` which makes the
```
```suggestion
        Helper function used after `:func:~sunpy.net.dataretriever.GenericClient.search` which makes the
```

abhijeetmanhas replied:
Done

Cadair commented on sunpy/net/dataretriever/client.py:
```diff
@@ -225,7 +176,63 @@ def _can_handle_query(cls, *query):
+            end = parse_time("{}/{}/{}".format(exdict['year'], 12, 31)) + almost_day
+        map_['Time'] = TimeRange(start, end)
+        map_['Start Time'] = start.strftime(TIME_FORMAT)
+        map_['End Time'] = end.strftime(TIME_FORMAT)
```
Why are we injecting the timerange and the start and end times, why not just one or the other?

abhijeetmanhas replied:
This was done so that we have `Time` as a Quantity in the table which can be used for inspection. And start time/endtime as strings which can be used for showing in the response table, so that results look similar as it was before for time.

If we inject only `Time`, we had to do that thing in `build_table`, and as I also had to preserve order of colnames in the table, it was relatively difficult to implement.

abhijeetmanhas replied:
Because scraper needs a `TimeRange` for generating urls.

abhijeetmanhas replied:
Also we would lose those milliseconds in endtime, if we use stringified time.

Cadair commented on sunpy/net/dataretriever/client.py:
```diff
@@ -1,63 +1,45 @@
 
 
-class QueryResponseBlock:
+class QueryResponseBlock(OrderedDict):
```
This PR feels like a good opportunity to get rid of this class? Or shall we do that later?

nabobalis replied:
I thought this was removed?

abhijeetmanhas replied:
I had to bring back `QueryResponseBlock` due to two reasons:
- All vso, jsoc, dataretriever individual records (rows in table) used to be classes, and to access their attributes, we have to do something like `row.url` instead of `row['ur']` in my implementation. So this ensured uniformity.
- A huge changes might have to be done in `database` tests, first by checking what type of response it is and whether to do `['url']` or `.url`. So it saved me from it.

nabobalis replied:
The question is, is it worth removing this class and just doing those annoying little changes everywhere?

abhijeetmanhas replied:
Removed in the latest commit, there weren't lot of changes.
Now the question is, should we keep the way to access fields from row data of QueryResponse (earlier QueryResponseBlock ) different for VSO and DR?
I think it would be rarely used. And after `:func:show` added, it won't be required.

nabobalis replied:
Different how?

abhijeetmanhas replied:
Like in VSO, the zeep client makes each block an object, so time is accessed like `block.time` or `getattr(block, 'time')`.

But I make them a dictionary, so it is accessed like `block['time']` or `block.get('time')`

nabobalis replied:
On the surface I would think we go for the latter, is it not how astropy.tables are indexed?

abhijeetmanhas replied:
Exactly, that's how we index cols in astropy table, like I'm doing in dict

nabobalis commented on examples/time_series/goes_hek_m25.py:
```diff
@@ -15,7 +15,7 @@
 ###############################################################################
 # Let's first grab GOES XRS data for a particular time of interest
-tr = TimeRange(['2011-06-07 04:00', '2011-06-07 12:00'])
+tr = TimeRange(['2011-06-07 00:00', '2011-06-07 12:00'])
```
What caused this change?

abhijeetmanhas replied:
The goes `xrs` had one day files, so my implementation defaults the start time of a record to `00:00` on that day.
When the start time is given `4:00` it isn't less than `0:00` so scraper doesn't return that file.

nabobalis replied:
Can we revert this now?

abhijeetmanhas replied:
yes, will revert.

abhijeetmanhas commented on sunpy/net/dataretriever/client.py:
```diff
@@ -137,86 +114,60 @@ class GenericClient(BaseClient):
-        This `map\\_` dictionary is passed through to the
-        `_get_url_for_timerange` method to get the URL results.
+        Returns a dictionary to validate the metadata of searched files using
+        query and registered values for Attrs for the respective client.
```
Reprhased

nabobalis commented on changelog/4321.feature.rst:
```diff
@@ -0,0 +1,14 @@
+Refactoring of `~sunpy.net.dataretriever` which lead to the following changes:
```
This should be split into multiple changelogs

abhijeetmanhas replied:
Done

nabobalis commented on changelog/4321.breaking.1.rst:
```diff
@@ -0,0 +1 @@
+Removed `:func:~sunpy.net.dataretriever.QueryResponse.create`.
```
Can you check if ":func:" should be outside or inside of the backticks?

abhijeetmanhas replied:
According to docs
```
python function :py:func:`package.name.function`
:py:meth:`package.name.ClassName.method`.
```
So ideally , we should use `:meth:` here. But we have used both meth and func across sunpy, in changelog and the rst docs. However, in this PR I'll use meth. And func should be outside those ticks, which also need to be changed.

nabobalis replied:
Can you use the correct one?

abhijeetmanhas replied:
Yes, I have used the correct one, in all changelogs now.

nabobalis commented on changelog/4321.breaking.2.rst:
```diff
@@ -0,0 +1 @@
+Removed `~sunpy.net.dataretriever.QueryResponseBlock` as `~sunpy.net.dataretriever.QueryResponse` now uses a ``list`` of `~collections.OrderedDict` for building the response table.
```
```suggestion
Removed `~sunpy.net.dataretriever.QueryResponseBlock` as `~sunpy.net.dataretriever.QueryResponse` now uses a `list` of `~collections.OrderedDict` for building the response table.
```

nabobalis commented on changelog/4321.feature.1.rst:
```diff
@@ -0,0 +1,3 @@
+Additions in `~sunpy.util.scraper` to support the refactoring of ``GenericClient``:
```
```suggestion
Additions in `~sunpy.util.scraper` to support the refactoring of ~sunpy.net.dataretriever.GenericClient:
```

hayesla commented on examples/time_series/goes_hek_m25.py:
```diff
@@ -15,7 +15,7 @@
 
 ###############################################################################
 # Let's first grab GOES XRS data for a particular time of interest
-tr = TimeRange(['2011-06-07 04:00', '2011-06-07 12:00'])
```
hmmmm this is an issue now I guess - will `tr = TimeRange(['2011-06-07 04:00', '2011-06-07 12:00'])` this now not return the file for this day? it really should - many people will search only for a flare time so this is important!

abhijeetmanhas replied:
This issue is due to all clients assuming that if a filename contains the timestamp of a file, 
For example goes contain `2011/6/7` in the url, so it start time is considered as `2011/6/7 00:00` since there's no info about `HHMMSS` and there's one file per day. 
Since scraper compares `start_time_in_attr<=start_time_extracted`, so if we do `2011/4/7 4:00`, so file for `2011/6/7` is not returned. What best thing we should be doing? 

One thing in my mind is compare inferred endtime of a file with the end time provided in the attr. If one of them is true, then we will return the file.

kakirastern replied:
I agree with @hayesla in that it is customary to return the file for the day being searched, otherwise it would be against the usual user expectations...

hayesla replied:
hmm is there a way to implement within goes.py? such that if the hh:mm is given it will automatically trim them to 00:00?

abhijeetmanhas replied:
@hayesla this is fixed now with some tweaks in scraper; by adding :func:`get_timerange_with_extractor` in scraper.

hayesla commented on sunpy/database/tests/test_database.py:
```diff
@@ -74,12 +74,11 @@ def fido_search_result():
     return Fido.search(
         net_attrs.Time("2012/1/1", "2012/1/2"),
-        net_attrs.Instrument('lyra') | net_attrs.Instrument('eve') |
+        net_attrs.Instrument('lyra') & net_attrs.Level.two | net_attrs.Instrument('eve') |
```
```suggestion
        net_attrs.Instrument('lyra') & net_attrs.Level(2) | net_attrs.Instrument('eve') |
```

hayesla commented on sunpy/database/tests/test_tables.py:
```diff
@@ -55,12 +55,11 @@ def fido_search_result():
     return Fido.search(
         net_attrs.Time("2012/1/1", "2012/1/2"),
-        net_attrs.Instrument('lyra') | net_attrs.Instrument('eve') |
+        net_attrs.Instrument('lyra') & net_attrs.Level.two | net_attrs.Instrument('eve') |
```
```suggestion
        net_attrs.Instrument('lyra') & net_attrs.Level(2) | net_attrs.Instrument('eve') |
```

hayesla commented on sunpy/net/dataretriever/sources/tests/test_goes_ud.py:
```diff
@@ -23,27 +23,29 @@ def LCClient():
 @pytest.mark.parametrize(
     "timerange,url_start,url_end",
-    [(TimeRange('1995/06/03', '1995/06/05'),
+    [(Time('1995/06/03', '1995/06/05'),
```
just a general comment - we should probably have some tests that also include %H%m to check this works too

abhijeetmanhas replied:
Added many tests with `%H%M` for goes.

hayesla commented:
This is great! 

I have one issue that needs to be addressed and fixed before this can be approved.
If for example you search over a timerange during a certain day doesn't return anything - 
for example:
```python
>>> result = Fido.search(a.Time('2013-10-28 01:00', '2013-10-28 03:00'), a.Instrument('XRS')) 
>>> print(result)
Results from 1 Provider:

0 Results from the XRSClient:
<No columns>

```

this is a bug that needs to be fixed!

hayesla commented:
this is similarly a problem is someone spans two days but not at 00:00

for example:
```python
>>> result = Fido.search(a.Time('2013-10-28 01:00', '2013-10-29 03:00'), a.Instrument('XRS')) 
>>> print(result)
Results from 1 Provider:

1 Results from the XRSClient:
     Start Time           End Time      Instrument  Physobs   Source Provider SatelliteNumber
------------------- ------------------- ---------- ---------- ------ -------- ---------------
2013-10-29 00:00:00 2013-10-29 23:59:59        XRS irradiance   NASA     SDAC              15
```
it only returns the second day.

Cadair commented on docs/guide/acquiring_data/fido.rst:
```diff
@@ -98,7 +85,7 @@ Searching for Data Using Fido
 For example::
 
-    >>> result = Fido.search(a.Time('2012/3/4', '2012/3/6'), a.Instrument.lyra) # doctest: +REMOTE_DATA
+    >>> result = Fido.search(a.Time('2012/3/4', '2012/3/6'), a.Instrument.lyra, a.Level.two) # doctest: +REMOTE_DATA
```
I think we need to decide which one we are doing and then do it everywhere. I think @nabobalis changed it everywhere applicable.

nabobalis replied:
Do whatever you want.

abhijeetmanhas replied:
Both will work by the way

hayesla replied:
either is fine - just a personal preference nothing else!

hayesla commented:
and just note this is an issue with all the clients it seems not just XRS

Cadair commented on sunpy/net/dataretriever/client.py:
```diff
@@ -225,7 +170,57 @@ def _can_handle_query(cls, *query):
+        map_['Start Time'] = start.strftime(TIME_FORMAT)
+        map_['End Time'] = end.strftime(TIME_FORMAT)
+        map_['Instrument'] = matchdict['Instrument'][0].upper()
+        if 'Physobs' in matchdict:
```
why only the if statement for physobs, what if my new client doesn't define a source, provider or even instrument? This whole method feels like it's pretty specialised for what exists at the moment rather than the design of the `GenericClient`.

Am I right in thinking that this is a converter from `cls.pattern` to a response table? Can we not generalise this by restricting (via docs) the possible names of parameters in `cls.pattern`?

Cadair replied:
In fact, couldn't we use the fields present in the pattern to know which things we need to process here?

abhijeetmanhas replied:
Some fields are not present in the `pattern`, like `Source`, `Instrument`, `Provider` etc. So we can't do that way. I'm using matchdict for that work. It uses `register_values` for getting those attrs. The problem is some things may not be even registered as attrs. (Like Time and Wavelength). So we need an individual compare after running loop over the things available from `cls.pattern`.

Yes you are right I shoudn't check only for `Physobs`. Infact I should check for all other attrs too.

Cadair replied:
This brings up an interesting thought about how we validate `register_values` and the pattern. If things aren't in the pattern they should have one value in `register_values` and then all the results take that value? If they have more than one value in `register_values` then they should be in the pattern?

Can you not use this logic to determine how to build the results? It would then be much more flexible. I wouldn't expect you to need to call out anything explicitly in this method other than maybe time?

abhijeetmanhas replied:
Yes, not assuming existence of specific attrs now. Applied that logic in `post_search_hook`

Cadair commented on sunpy/net/dataretriever/client.py:
```diff
@@ -225,7 +170,57 @@ def _can_handle_query(cls, *query):
+        """
+        map_ = OrderedDict()
+        almost_day = TimeDelta(1 * u.day - 1 * u.millisecond)
+        if 'month' in exdict and 'day' in exdict:
```
Now I think about it, the fact there is `"year", "month" and "day"` in this dict at all is weird.

abhijeetmanhas replied:
https://github.com/sunpy/sunpy/pull/4321#discussion_r463059042
https://github.com/sunpy/sunpy/pull/4321#discussion_r463059949
this all is for getting end times for a url. If file contains only `year` then end time is defaulted to end of that year.
If file contains only `y/m/d` then endtime is defaulted to end of the day,

It works for all files because none of the file have time in format `%j` or other format. In `SUVICLient` however the endtime was in the filename itself, so `post_search_hook` was overwritten.
All DR clients have daily files, except `SRS` has yearly file. But I think it is a bad assumption. Some scraper changes can be helpful to address it.

Cadair commented on sunpy/net/dataretriever/client.py:
```diff
@@ -225,7 +170,57 @@ def _can_handle_query(cls, *query):
+        return map_
 
     def _get_full_filenames(self, qres, filenames, path):
         """
```
This docstring is telling lies, it downloads nothing.

abhijeetmanhas replied:
Yes, updated.

Cadair commented on sunpy/net/dataretriever/client.py:
```diff
@@ -281,18 +267,22 @@ def search(self, *args, **kwargs):
+                                                matcher=matchdict)
+        metalist = []
+        for i in filesmeta:
+            map_ = self.post_search_hook(i, matchdict)
```
:x: 

*hisses at `map_`*

Cadair commented on sunpy/net/dataretriever/sources/goes.py:
```diff
@@ -222,184 +84,108 @@ class SUVIClient(GenericClient):
+    def search(self, *args, **kwargs):
+        supported_waves = [94, 131, 171, 195, 284, 304]*u.Angstrom
+        all_waves = []
+        baseurl1b, pattern1b, baseurl2, pattern2, matchdict = self.pre_search_hook(*args, **kwargs)
```
why even bother with a call to `pre_search_hook` in this client?

abhijeetmanhas replied:
You're right, not calling now.

Cadair commented on sunpy/net/dataretriever/sources/noaa.py:
```diff
@@ -32,55 +34,33 @@ class NOAAIndicesClient(GenericClient):
+
+    def search(self, *args, **kwargs):
+        map_ = self._get_match_dict(*args, **kwargs)
+        for key in map_:
+            if isinstance(map_[key], list):
+                map_[key] = map_[key][0]
```
I think I know what's happening here, but a comment would help.

abhijeetmanhas replied:
Added.

Cadair commented on sunpy/net/dataretriever/sources/noaa.py:
```diff
@@ -102,59 +82,33 @@ class NOAAPredictClient(GenericClient):
+
+    def search(self, *args, **kwargs):
+        map_ = self._get_match_dict(*args, **kwargs)
+        for key in map_:
+            if isinstance(map_[key], list):
+                map_[key] = map_[key][0]
```
same here.

Cadair commented on sunpy/net/dataretriever/sources/noaa.py:
```diff
@@ -200,6 +158,21 @@ def _get_url_for_timerange(self, timerange, **kwargs):
+            exdict1 = parse(extractor1, url)
+            exdict2 = parse(extractor2, url)
+            exdict = (exdict2 if exdict1 is None else exdict1).named
+            map_ = super().post_search_hook(exdict, matchdict)
```
why `super()` we haven't overriden it right?

abhijeetmanhas replied:
Actually the `SRSClient` is weird. It changes its pattern on the basis of year. So I first know the correct pattern and then use the `super().search` cause it saves me from adding scraper code.

Cadair replied:
My point  was `self.post_search_hook` should be fine.

abhijeetmanhas replied:
Oh ok, corrected.

Cadair commented on sunpy/net/dataretriever/client.py:
```diff
@@ -137,86 +101,67 @@ class GenericClient(BaseClient):
+    baseurl = None
+    pattern = None
+    required = {a.Time, a.Instrument}
+    optional = set()
```
I found [this](https://github.com/sunpy/sunpy/pull/4321/files#r449536787) old thread where it was pointed out that the main issue is that `Wavelength` (and I assume `Time`) isn't registered. This makes me think we should be registering them, even if we can't register values for them.

@nabobalis what do you think about that? Could we just pass `{a.Wavelength: '*'}` and everything just work?

nabobalis replied:
Try it and see?

nabobalis replied:
I am not sure I ever tested non-simple attrs with `*`

abhijeetmanhas replied:
Removed optional by adding Time and Wavelength in `register_values()`.

abhijeetmanhas commented on changelog/4321.feature.1.rst:
```diff
@@ -0,0 +1,3 @@
+Additions in `~sunpy.util.scraper` to support the refactoring of `~sunpy.net.dataretriever.GenericClient`:
+- :meth:`~sunpy.util.scraper.Scraper.findDatewith_extractor` which uses the parses the url using extractor to return its start time.
```
Yes,it was wrong.
Changed to
```- :meth:`~sunpy.util.scraper.Scraper.findDatewith_extractor` which parses the url using extractor to return its start time.
```

abhijeetmanhas commented on sunpy/net/dataretriever/client.py:
```diff
@@ -137,86 +101,67 @@ class GenericClient(BaseClient):
-    does this through the `map\\_` dictionary which represents the
-    `~sunpy.net.attrs` objects as a dictionary.
     """
+    baseurl = None
```
Done.

abhijeetmanhas commented on sunpy/net/dataretriever/client.py:
```diff
@@ -225,7 +170,57 @@ def _can_handle_query(cls, *query):
         class uses to dispatch queries to this Client.
         """
-        raise NotImplementedError
+        adict = cls.register_values()
```
Changed to `regattrs_dict`

abhijeetmanhas commented on sunpy/net/dataretriever/client.py:
```diff
@@ -137,86 +101,67 @@ class GenericClient(BaseClient):
     `~sunpy.net.dataretriever.client.GenericClient.fetch` the former generates a
```
Added some details of post and pre hooks. Also explained `baseurl` `pattern` etc.

abhijeetmanhas commented on sunpy/net/dataretriever/client.py:
```diff
@@ -137,86 +101,67 @@ class GenericClient(BaseClient):
+        Returns
+        -------
+        matchdict: `dict`
+            A dictionary having a `list` of all possible Attr values
```
It knows them all by `register_values()`. Only those which are registered in that method will be known to the client.

Cadair replied:
This seems flawed. The big change to GOES in particular was us being able to search all the available satellites, but because we are only search for ones which we have listed in `register_values` if a new satellite becomes operational we would need the user to upgrade sunpy before they could get results for that satellite.

This pulls us back to redesigning scraper doesn't it?

hayesla replied:
I think it's probably ok given that its somewhat of a rare occurrence that there'll be a new GOES satellite with science quality XRS data so I think this is ok to have it 'hard-coded' into the register_values. 

Like the more I think about this - people really will only want just GOES data from one instrument given that they are _supposed_ to be the same data (calibrated to be the same anyway). Those that are interested in using other GOES satellites (e.g. wanting to use GOES 13 when there is GOES 15) will could just use the `SatelliteNumber` attrs. The majority of the community will really just want the lightcurves from the most current instrument at the time of their observation. We could just have a nice example of how to do this etc. 

But as @abhijeetmanhas mentioned in the matrix chat - how would scraper work when the pattern is `../satellite_number/y/m/d...` which is the case for GOES 16 and 17 (rather than just listing all files under `../goes_data/fits/y/m/d..` like for previous data) you would have to register it somewhere times when this data is available and iterate to list the available data in the search right?

abhijeetmanhas replied:
yes, because else we won't be able to know the directory names where our scraper would search data for.

Cadair replied:
Conceptually I don't see why the scraper shouldn't be able to handle a pattern like that, list all directories, use the values. Bonus points for checking that the format is something like `{:d2}` or whatever.

abhijeetmanhas commented on sunpy/net/dataretriever/client.py:
```diff
@@ -225,7 +170,57 @@ def _can_handle_query(cls, *query):
+            An Ordered Dictionary which is used by `QueryResponse`
+            to show results.
+        """
+        map_ = OrderedDict()
```
changed to `rowdict` everywhere.

abhijeetmanhas commented:
Things which are yet to be resolved are
- Handling of time in `client.py`.
- Handling of wildcard attrs, optionaldict.

abhijeetmanhas commented on sunpy/net/dataretriever/client.py:
```diff
@@ -225,7 +170,57 @@ def _can_handle_query(cls, *query):
+        for key in adict:
+            all_vals = [i[0].lower() for i in adict[key]]
+            for x in query:
+                if isinstance(x, key) and str(x.value).lower() not in all_vals:
```
yes it will fail, the code assumes that the one who will write client won't add wildcard. I will workout how it can be supported here.

abhijeetmanhas replied:
https://github.com/sunpy/sunpy/pull/4321#discussion_r466726547

Cadair commented on changelog/4321.breaking.rst:
```diff
@@ -0,0 +1 @@
+Removed :meth:`~sunpy.net.dataretriever.GenericClient._makeargs`, :meth:`~sunpy.net.dataretriever.GenericClient._get_url_for_timerange`, :meth:`~sunpy.net.dataretriever.GenericClient._makeimap`, :meth:`~sunpy.net.dataretriever.GenericClient_get_time_for_url.`, :meth:`~sunpy.net.dataretriever.GenericClient._link`.
```
We don't need a changelog entry for private (`_`) methods.

abhijeetmanhas replied:
Ok, removed.

abhijeetmanhas commented on sunpy/net/dataretriever/client.py:
```diff
@@ -225,7 +170,57 @@ def _can_handle_query(cls, *query):
+        """
+        map_ = OrderedDict()
+        almost_day = TimeDelta(1 * u.day - 1 * u.millisecond)
+        if 'month' in exdict and 'day' in exdict:
+            start = parse_time("{}/{}/{}".format(exdict['year'], exdict['month'], exdict['day']))
+            end = start + almost_day
+        elif 'year' in exdict:
+            start = parse_time("{}/{}/{}".format(exdict['year'], 1, 1))
+            end = parse_time("{}/{}/{}".format(exdict['year'], 12, 31)) + almost_day
```
https://github.com/sunpy/sunpy/pull/4321#discussion_r464013973

abhijeetmanhas commented on sunpy/net/dataretriever/client.py:
```diff
@@ -225,7 +170,57 @@ def _can_handle_query(cls, *query):
+            to show results.
+        """
+        map_ = OrderedDict()
+        almost_day = TimeDelta(1 * u.day - 1 * u.millisecond)
```
Is it Ok if I add method to get timerange for a file in each client itself?

abhijeetmanhas replied:
This is fixed, we don't have defaulted timerange to one day in `client.py` now.

abhijeetmanhas commented on sunpy/util/scraper.py:
```diff
@@ -137,10 +137,24 @@ def _URL_followsPattern(self, url):
             return matches.end() == matches.endpos
         return False
 
+    def _findDatewith_extractor(self, url):
+        exdict = parse(self.extractor, url).named
+        datetypes = ['year', 'month', 'day']
+        timetypes = ['hour', 'minute', 'second']
+        dtlist = []
+        for d in datetypes:
+            dtlist.append(exdict.get(d, 1))
+        for t in timetypes:
+            dtlist.append(exdict.get(t, 0))
+        dt = datetime.datetime(*dtlist)
+        return Time(dt)
```
This is not a method which follows generalization.

abhijeetmanhas replied:
Much general method added now.

nabobalis commented on sunpy/net/dataretriever/sources/rhessi.py:
```diff
@@ -151,65 +151,23 @@ def get_observing_summary_dbase_file(time):
+        for url in self.get_observing_summary_filename(timerange):
+            exdict = parse(pattern, url).named
+            exdict['url'] = url
+            map_ = super().post_search_hook(exdict, matchdict)
+            metalist.append(map_)
```
Is `map_` a good name for this variable?

abhijeetmanhas replied:
Changed to `rowdict` now everywhere.

nabobalis commented on sunpy/net/dataretriever/sources/noaa.py:
```diff
@@ -223,7 +200,7 @@ def fetch(self, qres, path=None, error_callback=None, **kwargs):
             name = url.split('/')[-1]
 
             # temporary fix !!! coz All QRBs have same start_time values
-            day = Time(qre.time.start.strftime('%Y-%m-%d')) + TimeDelta(i*u.day)
+            day = Time(qre['Time'].start.strftime('%Y-%m-%d')) + TimeDelta(i*u.day)
```
What on earth is this comment?!

abhijeetmanhas replied:
Don't know what are `QRB` (may be queryresponse) here, but it probably mentions about the assumption that all file have `00:00` as start time values, which is used in `fetch` defined differently here to name the downloaded files.

nabobalis replied:
Can we remove this "temp" fix?

abhijeetmanhas replied:
I feel we should keep it, `SRSClient` is exceptional in terms of having its own `fetch()` method, so if there's generalization later done with `fetch`, this comment can be helpful.

nabobalis commented on .pre-commit-config.yaml:
```diff
@@ -66,13 +66,13 @@ repos:
         args: ['--count', '--select', 'E101,E11,E111,E112,E113,E121,E122,E123,E124,E125,E126,E127,E128,E129,E131,E133,E20,E211,E231,E241,E242,E251,E252,E26,E265,E266,E27,E301,E302,E303,E304,E305,E306,E401,E402,E502,E701,E711,E712,E713,E714,E722,E731,E901,E902,F822,F823,W191,W291,W292,W293,W391,W601,W602,W603,W604,W605,W690']
```
Why are these changes in the diff?

abhijeetmanhas replied:
I messed up in git, let me check.

nabobalis replied:
If you are having issues, it might be better to checkout the old version of this PR. Squash each commit and then rebase it.

abhijeetmanhas replied:
Resolved now.

hayesla commented on sunpy/net/dataretriever/sources/tests/test_eve.py:
```diff
@@ -134,29 +134,29 @@ def mock_query_object(LCClient):
     # Creating a Query Response Object
     start = '2016/1/1'
-    end = '2016/1/2'
+    end = '2016/1/1 23:59:59'
```
Just out of curiousity - why are all the end times for the tests (incl. ones for other instruments) all changed to 2016/01/01 23:59:59 rather than just 2016/01/02? - is this left over from the problem with scraper before?

abhijeetmanhas replied:
This is query response object to check the end time from the responses returned from table.
Earlier we defaulted it to end date for some clients and enddate - 1 microseconds for some. Now it is consistent to enddate -1 microseconds.

hayesla commented:
This looks great - and the problems with overlapping times and querying times on the same day now is fixed! 🎉

nabobalis commented:
> This is great!
> 
> I have one issue that needs to be addressed and fixed before this can be approved.
> If for example you search over a timerange during a certain day doesn't return anything -
> for example:
> 
> ```python
> >>> result = Fido.search(a.Time('2013-10-28 01:00', '2013-10-28 03:00'), a.Instrument('XRS')) 
> >>> print(result)
> Results from 1 Provider:
> 
> 0 Results from the XRSClient:
> <No columns>
> ```
> 
> this is a bug that needs to be fixed!

Can we add this as a test case?

nabobalis commented:
> Things which are yet to be resolved are
> 
> * Handling of time in `client.py`.
> * Handling of wildcard attrs, optionaldict.

Are both of these resolved?

abhijeetmanhas commented:
> > Things which are yet to be resolved are
> > 
> > * Handling of time in `client.py`.
> > * Handling of wildcard attrs, optionaldict.
> 
> Are both of these resolved?

yes

abhijeetmanhas commented:
> > This is great!
> > I have one issue that needs to be addressed and fixed before this can be approved.
> > If for example you search over a timerange during a certain day doesn't return anything -
> > for example:
> > ```python
> > >>> result = Fido.search(a.Time('2013-10-28 01:00', '2013-10-28 03:00'), a.Instrument('XRS')) 
> > >>> print(result)
> > Results from 1 Provider:
> > 
> > 0 Results from the XRSClient:
> > <No columns>
> > ```
> > 
> > 
> > this is a bug that needs to be fixed!
> 
> Can we add this as a test case?

added

Cadair reviewed (approved):
This looks complete, and it's holding up other stuff so let's do it!
